### DISCUSSION TITLE
Alternative Marketplace proposal

### DISCUSSION THREAD

**Discussion by Bulat-Ziganshin** [codex-storage/codex-research#101] [General]
TLDR: Two-step contracts with pre-contract stage and punishments for all sides if full contract wasn't signed at the end.

We have:
- 1 contract
- consisting of `N=M+K` slots. Note that slots are equal-sized, thus interchangeable
- overprovisioning: CN may choose to sign `O` pre-contracts for each slot to maximize chances that full contract will be signed

Once contract proposal was published, CN starts to get quotes from interested nodes. Once CN gets enough quotes, it choses `O*N` of them and signs pre-contracts with these nodes. These nodes, called PHN, stakes their collateral for full contract at this point. Slot number is assigned to each PHN at this point rather than selected by PHN itself.

Once pre-contract is signed (simultaneously with all PHN, just like the full contract), each PHNs downloads data from the assigned slot and publish PoR.

Once PoR for each slot is published, the full contract may be signed automatically. It includes for each slot the first node that published PoR for this slot, this node promoted from PHN to HN. Other PHN get their collateral back and got out of contract.

If final contract is successfully signed, CN pays each HN f.e. `3*O*D` coins, where D should be just enough to cover real download expenses of average network node. Since each node has `1/O` chances to get into contract, its average income from successful contracts will be `3*D`.

If after the amount of time specified in published contract, not all slots are filled by publishing PoRs - contract is automatically dropped. In this case, each involved node is penalized, f.e.
- CN pays `O*N*D` coins
- each PHN pays `D` coins, i.e. `O*N*D` total

Since PHN gets on average `3*D` coins for each successful contract and penalized by `D` coins on unsuccessful ones plus it spends equivalent of `D` coins downloading each dataset, it will not lose money if at least 50% of contracts are successful.

##

This scheme repels the following attacks:
- CN signs pre-contracts but don't allow to download all data in time - CN loses some money
- Attacker gets significant part of pre-contracts but don't plan to download data / publish PoR - his losses are proportional to the fraction of pre-contracts he got, as well as his impact on the contract fate
- Scammer signs pre-contract but don't downloads data or downloads only 99% and stops - he will be never paid and will lose money if it will prevent signing of contract

In other words, every node that impacts chances of contract being signed, in the case it's not signed, pays proportional to the impact they had.

Also, once a PHN downloaded slot data and published PoR, other nodes in this slot no more have incentive to continue download data and thus waste the network bandwidth.

##

The contract is 2-staged:
- at the first stage (pre-contract) each PHN stakes the full collateral required for full contract. It's obliged to sign full contract if CN will choose it. And this collateral is used, in particular, to punish this PHN if full contract will fail
- once POR for each slot was published, the full contract may be started automatically by the network (in particular because each PHN is already staked full collateral required for it).

If it (POR publishing) doesn't happen, the contract is cancelled, and each PHN as well as CN lose specified amount of money, automatically too. These means that at pre-contract stage all these nodes should reserve at least amount of money enough for this punishment, in the form of collateral

##

Only host that first filled the slot and thus included in full contract, gets reward. But reward is `3*O*D`, so on average each host gets reward of `3*D`, covering its expenses for all successful and failed contracts he was involved (as far as at least 50% contracts succeed).

This ensures that at average, host will be paid enough to cover its download costs, but at the same time, lazy hosts that don't download data or download only 99% of data and stopped, have no chances to be rewarded. That's important part of protection against scammers and attacks.

The real drawback of this scheme is that my computations assume that all hosts have equal chances in the race to fill slot. Actually, hosts with better network connection will win more, thus demotivating other hosts.

The problem of hosts with better network network connection winning more might be alleviated by the expanding window proposal (from the "Dispersal" part of the proposal: https://github.com/status-im/codex-research/pull/83/files?short_path=f30397b#diff-f30397b16d98efe2ed923f207136d3d653d8ba36bb8515ac2be43bbdf1c6981a)
